Mathlib/Data/Finset/Grade.lean

```diff
@@ -0,0 +1,160 @@
+  ⟨Covby.exists_finset_cons, by rintro ⟨a, ha, rfl⟩; exact covby_cons _⟩
+
+lemma _root_.Covby.card_finset (h : s ⋖ t) : s.card ⋖ t.card := (val_covby_val.2 h).card_multiset
+
+section DecidableEq
+variable [DecidableEq α]
+
+@[simp] lemma wcovby_insert (s : Finset α) (a : α) : s ⩿ insert a s := by simp [← coe_wcovby_coe]
+@[simp] lemma erase_wcovby (s : Finset α) (a : α) : s.erase a ⩿ s := by simp [← coe_wcovby_coe]
+
+lemma covby_insert (ha : a ∉ s) : s ⋖ insert a s :=
+  (wcovby_insert _ _).covby_of_lt $ ssubset_insert ha
+
+@[simp] lemma erase_covby (ha : a ∈ s) : s.erase a ⋖ s := ⟨erase_ssubset ha, (erase_wcovby _ _).2⟩
+
+lemma _root_.Covby.exists_finset_insert (h : s ⋖ t) : ∃ a ∉ s, insert a s = t := by
+  simpa using h.exists_finset_cons
+
+lemma _root_.Covby.exists_finset_erase (h : s ⋖ t) : ∃ a ∈ t, t.erase a = s := by
+  simpa only [← coe_inj, coe_erase] using h.finset_coe.exists_set_sdiff_singleton
+
+lemma covby_iff_exists_insert : s ⋖ t ↔ ∃ a ∉ s, insert a s = t := by
+  simp only [← coe_covby_coe, Set.covby_iff_exists_insert, ← coe_inj, coe_insert, mem_coe]
+
+lemma covby_iff_card_sdiff_eq_one : t ⋖ s ↔ t ⊆ s ∧ (s \ t).card = 1 := by
+  rw [covby_iff_exists_insert]
+  constructor
+  · rintro ⟨a, ha, rfl⟩
+    simp [*]
+  · simp_rw [card_eq_one]
+    rintro ⟨hts, a, ha⟩
+    refine ⟨a, (mem_sdiff.1 $ superset_of_eq ha $ mem_singleton_self _).2, ?_⟩
+    rw [insert_eq, ← ha, sdiff_union_of_subset hts]
+
+lemma covby_iff_exists_erase : s ⋖ t ↔ ∃ a ∈ t, t.erase a = s := by
+  simp only [← coe_covby_coe, Set.covby_iff_exists_sdiff_singleton, ← coe_inj, coe_erase, mem_coe]
+
+end DecidableEq
+
+@[simp] lemma isAtom_singleton (a : α) : IsAtom ({a} : Finset α) :=
+  ⟨singleton_ne_empty a, fun _ ↦ eq_empty_of_ssubset_singleton⟩
+
+protected lemma isAtom_iff : IsAtom s ↔ ∃ a, {a} = s :=
+  bot_covby_iff.symm.trans $ covby_iff_exists_cons.trans $ by simp
+
+section Fintype
+variable [Fintype α] [DecidableEq α]
+
+lemma isCoatom_compl_singleton (a : α) : IsCoatom ({a}ᶜ : Finset α) := (isAtom_singleton a).compl
+
+protected lemma isCoatom_iff : IsCoatom s ↔ ∃ a, {a}ᶜ = s := by
+  simp_rw [← isAtom_compl, Finset.isAtom_iff, compl_eq_iff_isCompl, eq_compl_iff_isCompl]
+
+end Fintype
+
+/-- Finsets are multiset-graded. This is not very meaningful mathematically but rather a handy way
+to record that the inclusion `Finset α ↪ Multiset α` preserves the covering relation. -/
+instance instGradeMinOrder_multiset : GradeMinOrder (Multiset α) (Finset α) where
+  grade := val
+  grade_strictMono := val_strictMono
+  covby_grade _ _ := Covby.finset_val
+  isMin_grade s hs := by rw [isMin_iff_eq_bot.1 hs]; exact isMin_bot
+
+@[simp] lemma grade_multiset_eq (s : Finset α) : grade (Multiset α) s = s.1 := rfl
+
+instance instGradeMinOrder_nat : GradeMinOrder ℕ (Finset α) where
+  grade := card
+  grade_strictMono := card_strictMono
+  covby_grade _ _ := Covby.card_finset
+  isMin_grade s hs := by rw [isMin_iff_eq_bot.1 hs]; exact isMin_bot
```

Comment:
Should this follow by composition in some way?

Reply:
Yes, although it's
* not practical
* gives worse defeq

We don't have any example of graded orders in mathlib yet so it's hard to tell what works and what doesn't yet.